Userland/Libraries/LibWeb/HTML/FormControlInfrastructure.cpp

```diff
@@ -40,4 +46,140 @@ WebIDL::ExceptionOr<Entry> create_entry(JS::Realm& realm, DeprecatedString const
+        else if (auto* input_element = dynamic_cast<HTML::HTMLInputElement*>(control.ptr()); input_element && (input_element->type() == "checkbox" || input_element->type() == "radio") && input_element->checked()) {
+            //  1. If the field element has a value attribute specified, then let value be the value of that attribute; otherwise, let value be the string "on".
+            auto value = input_element->value();
+            if (value.is_empty())
+                value = "on";
+
+            // 2. Create an entry with name and value, and append it to entry list.
+            TRY_OR_THROW_OOM(realm.vm(), form_data_entries.try_append(value));
+            TRY_OR_THROW_OOM(realm.vm(), entry_list.try_set(name, form_data_entries));
+        }
+        // 8. Otherwise, if the field element is an input element whose type attribute is in the File Upload state, then:
+        else if (input_element && input_element->type() == "file") {
+            // 1. If there are no selected files, then create an entry with name and a new File object with an empty name, application/octet-stream as type, and an empty body, and append it to entry list.
+            if (input_element->files()->length() == 0) {
+                FileAPI::FilePropertyBag options {};
+                options.type = "application/octet-stream";
+                auto file = TRY(FileAPI::File::create(realm, {}, "", options));
+                TRY_OR_THROW_OOM(realm.vm(), form_data_entries.try_append(JS::make_handle(*file)));
+                TRY_OR_THROW_OOM(realm.vm(), entry_list.try_set(name, form_data_entries));
+            }
+            // 2. Otherwise, for each file in selected files, create an entry with name and a File object representing the file, and append it to entry list.
+            else {
+                for (size_t i = 0; i < input_element->files()->length(); i++) {
+                    auto* file = input_element->files()->item(i);
+                    TRY_OR_THROW_OOM(realm.vm(), form_data_entries.try_append(JS::make_handle(const_cast<FileAPI::File*>(file))));
+                }
+                TRY_OR_THROW_OOM(realm.vm(), entry_list.try_set(name, form_data_entries));
+            }
+        }
+        // FIXME: 9. Otherwise, if the field element is an input element whose type attribute is in the Hidden state and name is an ASCII case-insensitive match for "_charset_":
+        // FIXME:    1. Let charset be the name of encoding if encoding is given, and "UTF-8" otherwise.
+        // FIXME:    2. Create an entry with name and charset, and append it to entry list.
+        // 10. Otherwise, create an entry with name and the value of the field element, and append it to entry list.
+        else if (input_element) {
+            TRY_OR_THROW_OOM(realm.vm(), form_data_entries.try_append(input_element->value()));
+            TRY_OR_THROW_OOM(realm.vm(), entry_list.try_set(name, form_data_entries));
+        }
```

Comment:
Can `input_element` be null at this point?

If yes, why does the spec say: `Otherwise,...` and if no could we add a `VERIFY` to check that.
(Of course it may fail because we're missing steps for now)